### ISSUE TITLE
Default Data Logging: Include the average time per epoch

### ISSUE THREAD

**Issue by lminer** [wandb/wandb#1547]
It would be nice to have a metric somewhere that shows how long each epoch takes. Then I could compare runs by their average speed. You do have duration right now, but I often run jobs for differing numbers of epochs so it isn't quite the same thing.

**Comment by cvphelps** [comment]
Thanks for the suggestion, this is an interesting idea. You could achieve this in a visualization by using a custom chart, pulling in the total run time, and dividing by the number of epochs. Then you could have a bar chart to compare average epoch time across runs in your project.

**Comment by vanpelt** [comment]
@lminer the trickiest part of this is wandb has no notion of an "epoch" built into it's logging.  We have a notion of "step" but users generally log metrics at multiple time steps per epoch.  We could look for a metric named "epoch" and extract the time between the increments but that's a little tricky.  

Since you likely want to able to sort of filter by this I would suggest the simplest solution would add this as it's own metrics in your calls to wandb.log, i.e. `wandb.log({"epoch_time": XXX})`.  For historical runs you could write a script using our python api to add this to a runs summary so it's available as a column in the table:

https://docs.wandb.com/ref/export-api/examples#update-metrics-for-a-run-after-run-finished

**Comment by lminer** [comment]
The `wandb.log` option would only work in a custom training loop so I guess I'd have to write my own custom callback that writes to tensorboard. If you have a notion of "step", would average time per step be a possible metric to log? That's essentially the same thing.

**Comment by noamsgl** [comment]
Any ways in 2023 to log time per epoch with a pytorch lightning stack?

**Comment by anmolmann** [comment]
Hey @lminer , we have added lots of features and implemented improvements in both our SDK and App. There could be a couple of ways to achieve your use-case:
1. [Use `define_metric`](https://docs.wandb.ai/guides/track/log/customize-logging-axes) to set a custom x axis. Custom x-axes are useful in contexts where you need to log different time steps in the past during training, asynchronously.
You should then be able to switch between your default x-axis for all your chart panels in the UI by clicking on the `->x` button as seen in the following snippet.
![image](https://github.com/wandb/wandb/assets/20401632/ef458b92-fef8-4908-98a0-21d35fb302b0)

2. You can [use `Expressions`](https://docs.wandb.ai/guides/app/features/panels/line-plot/reference#expression) in the UI and customize your x-axis expression in the UI itself.
![image](https://github.com/wandb/wandb/assets/20401632/1e3ead73-08a5-48c0-a8fa-81f75f42a86e)

Please let us know in case of further queries.

**Comment by anmolmann** [comment]
Hi @lminer , since we have not heard back from you we are going to close this request. If you would like to re-open the conversation, please let us know!
